torch/distributed/device_mesh.py

```diff
@@ -52,9 +54,9 @@ def __init__(self) -> None:
                 int, tuple[Optional[str], Optional[C10dBackend.Options]]
             ] = {}
             self.root_to_flatten_mapping: dict[DeviceMesh, dict[str, DeviceMesh]] = {}
-            # Record flatten mesh name to its mesh dim index in root mesh.
-            self.flatten_name_to_root_dims: dict[
-                DeviceMesh, dict[str, tuple[int, ...]]
+            # Record flatten mesh name to its flattened layout in root mesh.
+            self.flatten_name_to_root_layout: dict[
```

Comment:
Can I ask if you intend to keep these global (thread-local) mappings, or if the plan is to eventually drop them? I still don't understand exactly why we need them, but if they're slated for removal I won't worry about them too much.

Reply:
If this is legacy, help me understand why storing MeshLayout over mesh dim index is preferable? For legacy things I would have naively expected you to keep them exactly as is and route around the damage.

Reply:
Probably not in this PR, and I'm not sure if you have done this in the unflatten PR, but we should add some TODOs and description about why these code exist and not the same as the unflatten design. BC is mostly the reason.